File: src/agents/random.ts
import type { Agent } from './types';

/** Uniform random over legal actions. The floor baseline. */
export function makeRandom(): Agent {
  return {
    name: 'random',
    choose(_obs, actions, ctx) {
      return actions[Math.floor(ctx.rng() * actions.length)];
    },
  };
}

/**
 * Random, but never ends a turn while it still has actions to spend and never
 * throws cards away to seize. A meaningfully stronger floor than `random`,
 * and the default rollout policy for search agents.
 */
export function makeRandomPlus(): Agent {
  return {
    name: 'random+',
    choose(_obs, actions, ctx) {
      const useful = actions.filter((a) => a.t !== 'endTurn' && a.t !== 'seize');
      const pool = useful.length > 0 ? useful : actions;
      return pool[Math.floor(ctx.rng() * pool.length)];
    },
  };
}

File: rust/crates/arcs-agents/src/random.rs
//! The floor baselines, ported from `src/agents/random.ts`.

use arcs_engine::{Action, Observation, Rng};

use crate::agent::{Agent, AgentCtx};
use crate::rollout::useful_indices;

/// Uniform random over legal actions. The floor baseline.
pub struct Random;

impl Agent for Random {
    fn name(&self) -> &str {
        "random"
    }

    fn choose(&mut self, _obs: &Observation, legal: &[Action], ctx: &mut AgentCtx) -> usize {
        ctx.rng.gen_range(legal.len())
    }
}

/// Random, but never ends a turn while it still has actions to spend and
/// never throws cards away to seize. A meaningfully stronger floor than
/// [`Random`], and the default rollout policy for search agents.
pub struct RandomPlus;

impl Agent for RandomPlus {
    fn name(&self) -> &str {
        "random+"
    }

    fn choose(&mut self, _obs: &Observation, legal: &[Action], ctx: &mut AgentCtx) -> usize {
        let useful = useful_indices(legal);
        if useful.is_empty() {
            ctx.rng.gen_range(legal.len())
        } else {
            useful[ctx.rng.gen_range(useful.len())]
        }
    }
}
